### ISSUE TITLE
It doesn't render certain characters

### ISSUE THREAD

**Issue by LuisMayo** [LuisMayo/ace-attorney-bot#53]
Reported by an user
Original report

> Hi. I would like to report an issue on ace attorney bot. When rendering threads in Ukrainian, it skips Cyrillic letters і and ї. I think this has something to do with encoding which is not utf-8. I know you may not have time to fix this, but if someday you would like to improve your bot somehow, that is a great thing to do! Love your bot.

Added report for Chinese

> Please see https://twitter.com/aceCourtBot/status/1462362972450340864
> 
> Some simplified Chinese characters are rendered as solid dot (·)

Same with Emoji

**Comment by Symbian9** [comment]
Where is font asset, used for render text, is actually hosted?

Is it pixelmap font asset or is it generated in some way directly from code during render?

**Comment by LuisMayo** [comment]
Hi

As you have already discovered, the font used is in the assets folder.
About changing the font, it wouldn't do harm to change it. As long as the new font supports the same number of languages/glyphs the old one did.

The thing is, it's not really that important. This project is not and it's not willing to be commercial. And about licensing, this project is already in a weird place for the copyright law. Take into consideration that Phoenix & Co. are protected until copyright, which cannot be used even in free projects. So I doubt a font for non-comercial usages (which again, this project doesn't have) is the biggest of our problems.

An improvement is an improvement though, so I may accept another font with a better license.

BTW I have to change Igiari, some user sent me via Telegram a modified Igiari.ttf with better support from turkish that I haven't added yet.

**Comment by Symbian9** [comment]
> BTW I have to change Igiari, some user sent me via Telegram a modified Igiari.ttf with better support from turkish that I haven't added yet.

What about replace "Igiari" to "Galmuri"? It has almost all needed characters:

- https://github.com/quiple/galmuri (could be also added as sub-module to repo)

Any missed character/glyph would be easily added to "Galmuri" ;)

**Comment by Symbian9** [comment]
@LuisMayo, how to create pull with font replacement changes?

**Comment by LuisMayo** [comment]
Just drop the zip here, fonts are outside Git control.
Unless you're adding new fonts, in that case you should make changes in text.py so drop the zip file in the Pull Request that has the text.py change

**Comment by Symbian9** [comment]
> Just drop the zip here, fonts are outside Git control.

OK, Here is `Galmuri 11` font TTF-file:

- https://github.com/quiple/galmuri/raw/main/dist/Galmuri11.ttf

**Comment by LuisMayo** [comment]
I've now tried it. The problem is it's not exactly the Ace Attorney font, which I enjoy more.

However it's a really good pixel font, with more charsets support than Igiari, so it's now the second font in the priority check (fonts get checked one by one until one perfect font has been found, Galmuri is the second to be texted, after Igiari).
Thanks for the font.

And talking about the issue at hand Ukranian should be a bit better now

**Comment by Symbian9** [comment]
> And talking about the issue at hand Ukranian should be a bit better now

Great news! Thanks.

JFTR, Does `@aceCourtBot` does not work at this time/today again?

- https://twitter.com/aceCourtBot/status/1485919939101970432

Because I just tried and there are no render yet:

- https://twitter.com/app4soft/status/1485944166861479936

**Comment by LuisMayo** [comment]
You're just in queue, I'm doing some experiments with the queue so it may be longer than expected!

**Comment by saschanaz** [comment]
I tried the new [Noto Emoji](https://fonts.google.com/noto/specimen/Noto+Emoji) and it kinda works, except it overflows from the line:

![image](https://user-images.githubusercontent.com/3396686/166139705-8357dcc4-ce86-4c3a-b202-7adae3239be2.png)

**Comment by LuisMayo** [comment]
Yes, you're right, that way we can set an specific size for the emoji font so it doesn't overflow while the rest of the text is in another font

**Comment by saschanaz** [comment]
So I found two issues for emojis so far.

1. `[:frame]` does not work for emojis as an emoji includes several unicode characters. Some library like https://pypi.org/project/grapheme/ should be used to properly render an emoji without breaking it down, but even that library does not support the latest emojis as it's not being maintained well.
1. [Pillow does not support font fallback](https://github.com/python-pillow/Pillow/issues/4808). There could be a handwritten way specifically for emojis where the library would put emojis in separate `AnimText` objects as it already does for newlines. (An emoji character can be detected by checking it against Noto Emoji)

**Comment by LuisMayo** [comment]
Yah, I'm afraid about the non fallback from pillow. That's thy I had to create the "compute best font" by myself.

Maybe a PR to the Pillow is warranted
